compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/cfa/CfgTraverser.kt

```diff
@@ -14,49 +25,159 @@ enum class TraverseDirection {
 fun <D> ControlFlowGraph.traverse(
     direction: TraverseDirection,
     visitor: ControlFlowGraphVisitor<*, D>,
-    data: D
+    data: D,
+    visitedSymbols: List<AbstractFirBasedSymbol<*>> = listOf(),
+    onNestingCall: (
+        TraverseDirection,
+        ControlFlowGraphVisitor<*, D>,
+        D,
+        List<AbstractFirBasedSymbol<*>>,
+        CFGNode<*>
+    ) -> Unit = { _, _, _, _, _ -> }
 ) {
     for (node in getNodesInOrder(direction)) {
         node.accept(visitor, data)
-        (node as? CFGNodeWithCfgOwner<*>)?.subGraphs?.forEach { it.traverse(direction, visitor, data) }
+        (node as? CFGNodeWithCfgOwner<*>)?.subGraphs?.forEach { it.traverse(direction, visitor, data, visitedSymbols, onNestingCall) }
+
+        // if this node is call node, onNestingCall is calling
+        if (node is FunctionCallNode || node is QualifiedAccessNode) {
+            onNestingCall(direction, visitor, data, visitedSymbols, node)
+        }
     }
 }
 
 fun ControlFlowGraph.traverse(
+    direction: TraverseDirection,
+    visitor: ControlFlowGraphVisitorVoid,
+    onNestingCall: (
+        TraverseDirection,
+        ControlFlowGraphVisitor<*, *>,
+        Any?,
+        List<AbstractFirBasedSymbol<*>>,
+        CFGNode<*>
+    ) -> Unit = { _, _, _, _, _ -> }
+) {
+    traverse(direction, visitor, null, onNestingCall = onNestingCall)
+}
+
+fun <D> ControlFlowGraph.traverseWithFunctionNesting(
```

Comment:
I'd think about name like `traverseInterprocedurally` or `traverseWithStepIntoFunctions`. Not quite sure about the best name here.